src/wallet/mod.rs

```diff
@@ -1689,16 +1689,75 @@ where
         let progress = progress.unwrap_or_else(|| Box::new(NoopProgress));
 
         let run_setup = self.ensure_addresses_cached(CACHE_ADDR_BATCH_SIZE)?;
-
         debug!("run_setup: {}", run_setup);
+
         // TODO: what if i generate an address first and cache some addresses?
         // TODO: we should sync if generating an address triggers a new batch to be stored
-        if run_setup {
-            maybe_await!(
-                blockchain.wallet_setup(self.database.borrow_mut().deref_mut(), progress,)
-            )?;
+
+        // first run, we record progress...
+        let mut sync_res = if run_setup {
+            maybe_await!(blockchain.wallet_setup(self.database.borrow_mut().deref_mut(), progress))
         } else {
-            maybe_await!(blockchain.wallet_sync(self.database.borrow_mut().deref_mut(), progress,))?;
+            maybe_await!(blockchain.wallet_sync(self.database.borrow_mut().deref_mut(), progress))
+        };
+
+        // We need to ensure descriptor is derivable to fullfil "missing cache", otherwise we will
+        // end up with an infinite loop
+        let is_deriveable = self.descriptor.is_deriveable()
+            && (self.change_descriptor.is_none()
+                || self.change_descriptor.as_ref().unwrap().is_deriveable());
+
+        if is_deriveable {
+            // Restrict max rounds in case of faulty "missing cache" implementation by blockchain
+            const MAX_ROUNDS: usize = 100;
+
+            for _ in 0..MAX_ROUNDS {
+                // If the error is the special `MissingCachedScripts` error, we return the number of
+                // scripts we should ensure cached.
+                // On any other error, we should return the error.
+                // On no error, we say `to_cache` is 0.
+                let to_cache = sync_res.map_or_else(
+                    |e| match e {
+                        Error::MissingCachedScripts(inner) => {
+                            // each call to `WalletSync` is expensive, maximize on scripts to search for
+                            println!("Wallet::sync => missing_count: {}", inner.missing_count);
+                            Ok(std::cmp::max(
+                                inner.missing_count as u32,
+                                CACHE_ADDR_BATCH_SIZE,
+                            ))
+                        }
+                        _ => Err(e),
+                    },
+                    |_| Ok(0_u32),
+                )?;
+
+                if to_cache == 0 {
+                    break;
+                }
+
+                // cache and try again
+                for keychain in [KeychainKind::External, KeychainKind::Internal] {
+                    let (_, keychain) = self._get_descriptor_for_keychain(keychain);
+                    let from = self
+                        .database
+                        .borrow()
+                        .iter_script_pubkeys(Some(keychain))?
+                        .len() as _;
```

Comment:
I would try to use the returned `last_count` value from the `MissingCachedScripts` error, to avoid this lookup which could be relatively slow in case the database has a lot of scripts.